Remove beginIndex() usage

permcody opened the pull request:
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

permcody committed:
Remove beginIndex() usage from framework

refs #13074

permcody committed:
Remove beingINdex() usage from modules

refs #13074

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/311594/) on a9b67af wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13143/site)

This comment will be updated on new commits.

andrsd commented on framework/src/transfers/MultiAppScalarToAuxScalarTransfer.C:
```diff
@@ -80,7 +80,7 @@ MultiAppScalarToAuxScalarTransfer::execute()
                        "MultiAppScalarToAuxScalarTransfer!");
 
-          for (auto j = beginIndex(from_values); j < from_values.size(); ++j)
+          for (unsigned int j = 0; j < from_values.size(); ++j)
```
Why can't you use `MooseIndex` here?

andrsd commented on framework/src/transfers/MultiAppScalarToAuxScalarTransfer.C:
```diff
@@ -120,7 +120,7 @@ MultiAppScalarToAuxScalarTransfer::execute()
                        "MultiAppScalarToAuxScalarTransfer!");
 
-          for (auto j = beginIndex(from_values); j < from_values.size(); ++j)
+          for (unsigned int j = 0; j < from_values.size(); ++j)
```
And here as well?

permcody replied:
Damn you noticed! I need to look at this more. This is a MooseArray and it failed to work with `MooseIndex`. I believe the issue is that it defines `size()`, which has something to do with the wrong template type being selected. I plan to look at this a little more as I'd like to make a sweep through the framework for all loops that start with `for (unsigned` and look for more candidates to change. This can probably go in as is for now, but it's not urgent if you'd prefer it be fixed first. I did check and the right hardcoded type for looping over a `MooseArray` is indeed `unsigned int`.

lindsayad committed:
Use declval for SFINAE for types with size method

This enables us to accept types that don't have default constructors like MooseArray

permcody commented:
Nicely done @lindsayad!

friedmud reviewed (approved):
👍
